jwst/pipeline/calwebb_image3.py

```diff
@@ -69,36 +66,41 @@ def process(self, input_data):
+        # Only load science members from input ASN;
+        # background and target-acq members are not needed.
+        if isinstance(input_data, ModelLibrary):
+            input_models = input_data
+        else:
+            input_models = ModelLibrary(input_data, asn_exptypes=['science'], on_disk=self.on_disk)
+
+        if self.output_file is None:
             # If input is an association, set the output to the product name.
-            if self.output_file is None:
-                try:
-                    self.output_file = input_models.meta.asn_table.products[0].name
-                except (AttributeError, IndexError):
-                    pass
-
-            # Check if input is single or multiple exposures
-            try:
-                has_groups = len(input_models.group_names) >= 1
-            except (AttributeError, TypeError, KeyError):
-                has_groups = False
-
-            if isinstance(input_models, ModelContainer) and has_groups:
-                if is_moving_target(input_models):
-                    input_models = self.assign_mtwcs(input_models)
-                else:
-                    input_models = self.tweakreg(input_models)
-
-                input_models = self.skymatch(input_models)
-                input_models = self.outlier_detection(input_models)
-
-            elif self.skymatch.skymethod == 'match':
-                self.log.warning("Turning 'skymatch' step off for a single "
-                                 "input image when 'skymethod' is 'match'")
-
-            else:
-                input_models = self.skymatch(input_models)
-
-            result = self.resample(input_models)
-            if isinstance(result, datamodels.ImageModel) and result.meta.cal_step.resample == 'COMPLETE':
-                self.source_catalog(result)
+            self.output_file = input_models.asn["products"][0]["name"]
+
+        with input_models:
+            model = input_models.borrow(0)
+            is_moving = is_moving_target(model)
+            input_models.shelve(model, 0, modify=False)
+        if is_moving:
+            input_models = self.assign_mtwcs(input_models)
+        else:
+            input_models = self.tweakreg(input_models)
+
+        input_models = self.skymatch(input_models)
+        input_models = self.outlier_detection(input_models)
+
+        # elif self.skymatch.skymethod == 'match':
+        #     self.log.warning("Turning 'skymatch' step off for a single "
+        #                      "input image when 'skymethod' is 'match'")
+
+        # else:
+        #     # FIXME: here input_models is a DataModel, passing
+        #     # that to skymatch would cause an error when it tries to call
+        #     # ModelContainer(DataModel). This can be seen by running
+        #     # strun calwebb_image3 any_cal.fits --steps.skymatch.method=local
+        #     input_models = self.skymatch(input_models)
```

Comment:
Where did this comment come from?  I don't remember writing this, and also don't understand it exactly, but it doesn't exist on master. @braingram was this part of your draft initially that I took over?

Reply:
Based on the docstring and [docs](https://jwst-pipeline.readthedocs.io/en/latest/jwst/pipeline/calwebb_image3.html#inputs) the image 3 pipeline can accept a single datamodel/filename input.

I'm not sure if this is tested. On main right now running:
```
strun calwebb_image3 any_cal.fits --steps.skymatch.skymethod=local
```
fails with:
```
TypeError: Input <...> is not a list of JwstDataModels or an ASN file
```
since the `elif` fails (due to `skymethod!=match`) and the `else` is executed and `skymatch` doesn't accept a single datamodel as an input.

On main, `strun calwebb_image3 <some_calfile>.fits` does run successfully.

With this PR I get `OSError: Cannot read ASN file.` (with or without `skymethod=local`) presumably where the pipeline attempts to convert the single datamodel filename to a `ModelLibrary`.

I guess one question is, is it ok to drop support for this untested mode (running with a single file)?

Reply:
We discussed this at stand-up. Although @tapastro was surprised `calwebb_image3` even ran with a single datamodel, we decided it should be easy enough to support this by just passing a one-member list into ModelLibrary.  I will try this

Reply:
Seems to work without any issues

Reply:
This was a little more complicated than I anticipated.  In short, if the input is an open datamodel, it's easy to just pass it as a single-member list.  However if the input is a filename, it's necessary to figure out whether the input is an association or a single file.  I did this with a `try...except` which seems to work.  I put the logic back in that says which steps should and shouldn't be run when the input is a single model.  Hopefully that means there have been no changes, but since the mode is untested, who knows?  I did at least add a unit test that ensures the pipeline runs to completion on a single `.cal` image filename.

Reply:
Thanks! This looks to be addressed in the updated code.